server/src/main/java/org/elasticsearch/cluster/coordination/ElectionStrategy.java

```diff
@@ -105,10 +110,14 @@ public void beforeCommit(long term, long version, ActionListener<Void> listener)
         listener.onResponse(null);
     }
 
-    public boolean nodeMayWinElection(ClusterState lastAcceptedState, DiscoveryNode node) {
+    public NodeEligibility nodeMayWinElection(ClusterState lastAcceptedState, DiscoveryNode node) {
         final String nodeId = node.getId();
-        return lastAcceptedState.getLastCommittedConfiguration().getNodeIds().contains(nodeId)
+        if (lastAcceptedState.getLastCommittedConfiguration().getNodeIds().contains(nodeId)
             || lastAcceptedState.getLastAcceptedConfiguration().getNodeIds().contains(nodeId)
-            || lastAcceptedState.getVotingConfigExclusions().stream().noneMatch(vce -> vce.getNodeId().equals(nodeId));
+            || lastAcceptedState.getVotingConfigExclusions().stream().noneMatch(vce -> vce.getNodeId().equals(nodeId))) {
+            return new NodeEligibility(true, "");
+        }
+
+        return new NodeEligibility(false, "node is ineligible for election, not a voting node in the voting configuration");
```

Comment:
Not sure how often is this executed. Is there a value in converting return values to constants to avoid extra allocations?